File: 2023/day1.py
from util import parse_file_line_by_line


def caluclate(numbers):
    return sum(numbers)


def get_value(line):
    first = None
    last = None
    for char in line:
        try:
            first = int(char)
            first = char
            break
        except:
            continue
    for char in line[::-1]:
        try:
            last = int(char)
            last = char
            break
        except:
            continue
    return int(first + last)


str_to_int_map = {
    "one": "1",
    "two": "2",
    "three": "3",
    "four": "4",
    "five": "5",
    "six": "6",
    "seven": "7",
    "eight": "8",
    "nine": "9",
}


def get_first_number_as_char(chars):
    number = None
    for i, char in enumerate(chars):
        try:
            number = int(char)
            number = char
            break
        except:
            pass
        for word_len in range(3, 6):
            if chars[i : i + word_len] in str_to_int_map:
                number = str_to_int_map[chars[i : i + word_len]]
                break
        if number:
            break
    return number


def get_last_number_as_char(chars):
    number = None
    for i, char in enumerate(chars):
        try:
            number = int(char)
            number = char
            break
        except:
            pass
        for word_len in range(3, 6):
            if i + word_len >= len(chars):
                continue
            if "".join(reversed(chars[i : i + word_len])) in str_to_int_map:
                number = str_to_int_map["".join(reversed(chars[i : i + word_len]))]
                break
        if number:
            break
    return number


def get_value_2(line):
    first = get_first_number_as_char(line)
    last = get_last_number_as_char(line[::-1])
    return int(first + last)


if __name__ == "__main__":
    print(sum(parse_file_line_by_line(1, 1, get_value)))
    print(sum(parse_file_line_by_line(1, 2, get_value_2)))

File: 2023/Day1.java
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Stream;


class Day1{
    private static final Map<String, Integer> DIGIT_NAMES = Map.of(
        "one", 1,
        "two", 2,
        "three", 3,
        "four", 4,
        "five", 5,
        "six", 6,
        "seven", 7,
        "eight", 8,
        "nine", 9
    );
    public static void main(String[] args) throws IOException {
        
        Path path = Paths.get("inputs/1.txt");
        Stream<String> lines = Files.newBufferedReader(path).lines();
        Integer part1 = lines.map(Day1::firstLastDigit).mapToInt(l -> l.getFirst()*10+l.getLast()).sum();
        System.out.println(part1);

        lines = Files.newBufferedReader(path).lines();
        Integer part2 = lines.map(Day1::firstLastDigitOrWord).mapToInt(l -> l.getFirst()*10+l.getLast()).sum();
        System.out.println(part2);
    }

    private static List<Integer> firstLastDigit(String line) {
        List<Integer> result = new ArrayList<>();
        for (int i=0; i < line.length(); ++i) {
            char c = line.charAt(i);
            if ('1' <= c && c <= '9') {
                result.add(c - '0');
            }
        }
        return result;
    }

    private static List<Integer> firstLastDigitOrWord(String line) {
        List<Integer> result = new ArrayList<>();
        for (int i=0; i < line.length(); ++i) {
            char c = line.charAt(i);
            if ('1' <= c && c <= '9') {
                result.add(c - '0');
            }
            for (int j = i; j <= i+6; j++){
                if (j <= line.length() && DIGIT_NAMES.containsKey(line.substring(i,j))) {
                    result.add(DIGIT_NAMES.get(line.substring(i,j)));
                }

            }
        }
        return result;
    }

}